sdk/storage/src/cloud_location.rs

```diff
@@ -44,28 +108,127 @@ impl CloudLocation {
+                            credentials: credentials.clone(),
+                        }
+                        .url(service_type),
+                        AZURE_CHINA_CLOUD => CloudLocation::China {
+                            account: account.clone(),
+                            credentials: credentials.clone(),
+                        }
+                        .url(service_type),
+                        AZURE_GERMAN_CLOUD => CloudLocation::GermanCloud {
+                            account: account.clone(),
+                            credentials: credentials.clone(),
+                        }
+                        .url(service_type),
+                        _ => {
+                            todo!()
+                        }
+                    };
+                } else if let Some(fallback) = fallback {
+                    return fallback.url(service_type);
+                } else {
+                    // Default to PROD
+                    return CloudLocation::Public {
+                        account: account.clone(),
+                        credentials: credentials.clone(),
+                    }
+                    .url(service_type);
+                }
+            }
         };
         Ok(url::Url::parse(&url)?)
     }
 
+    /// Returns the storage credentials for this cloud location,
+    ///
     pub fn credentials(&self) -> &StorageCredentials {
         match self {
-            CloudLocation::Public { credentials, .. } => credentials,
-            CloudLocation::China { credentials, .. } => credentials,
+            CloudLocation::Public { credentials, .. }
+            | CloudLocation::China { credentials, .. }
+            | CloudLocation::USGov { credentials, .. }
+            | CloudLocation::GermanCloud { credentials, .. }
+            | CloudLocation::Custom { credentials, .. }
+            | CloudLocation::AutoDetect { credentials, .. } => credentials,
             CloudLocation::Emulator { .. } => &EMULATOR_CREDENTIALS,
-            CloudLocation::Custom { credentials, .. } => credentials,
+        }
+    }
+
+    /// Finds the cloud name, first by environment variable, then by parsing the current user's $HOME/.azure/config file
+    ///
+    fn find_cloud_name() -> Option<String> {
+        if let Ok(name) = std::env::var("AZURE_CLOUD_NAME") {
+            Some(name)
+        } else if let Ok(home_dir) = std::env::var("HOME") {
+            if let Some(config) = PathBuf::from(home_dir)
+                .join(".azure/config")
+                .canonicalize()
+                .ok()
+                .and_then(|config| File::open(config).ok())
+            {
+                let mut lines = BufReader::new(config).lines();
+
+                while let Some(Ok(line)) = lines.next() {
+                    // Alternatively, import serde_toml for parsing this file, but probably better off doing
+                    // that by creating a struct dedicated to managing the config file
```

Comment:
we should either do this or remove this comment. I'm unsure if the dependency is worth it if we can keep things this short.